storage/cluster/database.go

```diff
@@ -0,0 +1,189 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cluster
+
+import (
+	"sync"
+
+	"github.com/m3db/m3db/sharding"
+	"github.com/m3db/m3db/storage"
+	"github.com/m3db/m3db/storage/namespace"
+	"github.com/m3db/m3db/topology"
+
+	"github.com/m3db/m3x/log"
+)
+
+var (
+	// newStorageDatabase is the injected constructor to construct a database,
+	// useful for replacing which database constructor is called in tests
+	newStorageDatabase newStorageDatabaseFn = storage.NewDatabase
+)
+
+type newStorageDatabaseFn func(
+	namespaces []namespace.Metadata,
+	shardSet sharding.ShardSet,
+	opts storage.Options,
+) (storage.Database, error)
+
+type clusterDB struct {
+	storage.Database
+
+	log    xlog.Logger
+	hostID string
+	watch  topology.MapWatch
+
+	watchMutex sync.Mutex
+	watching   bool
+	doneCh     chan struct{}
+	closedCh   chan struct{}
+}
+
+// NewDatabase creates a new clustered time series database
+func NewDatabase(
+	namespaces []namespace.Metadata,
+	hostID string,
+	topoInit topology.Initializer,
+	opts storage.Options,
+) (Database, error) {
+	log := opts.InstrumentOptions().Logger()
+	topo, err := topoInit.Init()
+	if err != nil {
+		return nil, err
+	}
+
+	watch, err := topo.Watch()
+	if err != nil {
+		return nil, err
+	}
+
+	// Wait for the topology to be available
+	<-watch.C()
```

Comment:
yeah my bad, waiting on the watch is safe as a new watch will get a initial push